qiskit_ode/signals/signals.py

```diff
@@ -327,202 +433,696 @@ def add_samples(self, start_sample: int, samples: List):
+            # convert to list if tuple
+            if type(sublist) == tuple:
+                sublist = list(sublist)
+        else:
+            # handle slices or singletons
+            sublist = operator.itemgetter(idx)(self.components)
+
+        # at this point sublist should either be a single Signal, or a list of Signals
+        if isinstance(sublist, list):
+            return self.__class__(sublist)
+        else:
+            return sublist
+
+    def conjugate(self) -> "SignalCollection":
+        """Return the conjugation of this collection."""
+        return self.__class__([sig.conjugate() for sig in self.components])
 
-        \Omega(t)*exp(2\pi i \nu t)
 
-    multiplication of two signals implements
+class SignalSum(SignalCollection, Signal):
+    r"""Represents a sum of ``Signal`` objects:
 
     .. math::
+        s_1(t) + \dots + s_k(t)
 
-        \Omega_1(t)*\Omega_2(t)*exp(2\pi i (\nu_1+\nu_2) t)
+    For basic evaluation, this class behaves in the same manner as ``Signal``:
 
+        - ``__call__`` evaluates the sum.
+        - ``complex_value`` evaluates the sum of the complex values of the individual summands.
 
-    Args:
-        sig1: A child of base signal or a constant.
-        sig2: A child of base signal or a constant.
+    Attributes ``carrier_freq`` and ``phase`` here correspond to an ``Array`` of
+    frequencies/phases for each term in the sum, and the ``envelope`` method returns an
+    ``Array`` of the envelopes for each summand.
 
-    Returns:
-        signal: The type will depend on the given base class.
+    Internally, the signals are stored as a list in the ``components`` attribute, which can
+    be accessed via direct subscripting of the object.
     """
 
-    # ensure both arguments are signals
-    if isinstance(sig1, (int, float, complex)):
-        sig1 = Constant(sig1)
+    def __init__(self, *signals, name: Optional[str] = None):
+        r"""Initialize with a list of Signal objects through ``args``\.
 
-    if isinstance(sig2, (int, float, complex)):
-        sig2 = Constant(sig2)
+        Args:
+            signals: ``Signal`` subclass objects.
+            name: Name of the sum.
 
-    return Signal(
-        lambda t: sig1.envelope_value(t) * sig2.envelope_value(t),
-        sig1.carrier_freq + sig2.carrier_freq,
-        sig1.phase + sig2.phase,
-    )
+        Raises:
+            QiskitError: If ``signals`` are not subclasses of ``Signal``\.
+        """
+        self._name = name
 
+        components = []
+        for sig in signals:
+            if isinstance(sig, list):
+                sig = SignalSum(*sig)
+
+            if isinstance(sig, SignalSum):
+                components += sig.components
+            elif isinstance(sig, Signal):
+                components.append(sig)
+            elif isinstance(sig, (int, float, complex)) or (
+                isinstance(sig, Array) and sig.ndim == 0
+            ):
+                components.append(Signal(sig))
```

Comment:
Seems like `Signal` should accept `Array` as input too? See my comment above.

Reply:
See response to above comment.